smartsim/settings/dispatch.py

```diff
@@ -0,0 +1,169 @@
+        self,
+        args: type[LaunchArgBuilder[_T]],
+        *,
+        to_launcher: type[LauncherLike[_T]],
+        allow_overwrite: bool = ...,
+    ) -> None: ...
+    def dispatch(
+        self,
+        args: type[LaunchArgBuilder[_T]] | None = None,
+        *,
+        to_launcher: type[LauncherLike[_T]],
+        allow_overwrite: bool = False,
+    ) -> t.Callable[[type[LaunchArgBuilder[_T]]], type[LaunchArgBuilder[_T]]] | None:
+        """A type safe way to add a mapping of settings builder to launcher to
+        handle the settings at launch time.
+        """
+
+        def register(
+            args_: type[LaunchArgBuilder[_T]], /
+        ) -> type[LaunchArgBuilder[_T]]:
+            if args_ in self._dispatch_registry and not allow_overwrite:
+                launcher_type = self._dispatch_registry[args_]
+                raise TypeError(
+                    f"{args_.__name__} has already been registered to be "
+                    f"launched with {launcher_type}"
+                )
+            self._dispatch_registry[args_] = to_launcher
+            return args_
+
+        if args is not None:
+            register(args)
+            return None
+        return register
+
+    def get_launcher_for(
+        self, args: LaunchArgBuilder[_T] | type[LaunchArgBuilder[_T]], /
+    ) -> type[LauncherLike[_T]]:
+        """Find a type of launcher that is registered as being able to launch
+        the output of the provided builder
+        """
+        if not isinstance(args, type):
+            args = type(args)
+        launcher_type = self._dispatch_registry.get(args, None)
+        if launcher_type is None:
+            raise TypeError(
+                f"{type(self).__name__} {self} has no launcher type to "
+                f"dispatch to for argument builder of type {args}"
+            )
+        # Note the sleight-of-hand here: we are secretly casting a type of
+        # `LauncherLike[Any]` to `LauncherLike[_T]`. This is safe to do if all
+        # entries in the mapping were added using a type safe method (e.g.
+        # `Dispatcher.dispatch`), but if a user were to supply a custom
+        # dispatch registry or otherwise modify the registry THIS IS NOT
+        # NECESSARILY 100% TYPE SAFE!!
+        return launcher_type
+
+
+default_dispatcher: t.Final = Dispatcher()
+
+
+# >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
+# TODO: move these to a common module under `smartsim._core.launcher`
+# -----------------------------------------------------------------------------
+
+
+def create_job_id() -> LaunchedJobID:
+    return LaunchedJobID(str(uuid.uuid4()))
+
+
+class LauncherLike(t.Protocol[_T_contra]):
+    def start(self, launchable: _T_contra) -> LaunchedJobID: ...
+    @classmethod
+    def create(cls, exp: Experiment) -> Self: ...
+
+
+class ShellLauncher:
+    """Mock launcher for launching/tracking simple shell commands"""
+
+    def __init__(self) -> None:
```

Comment:
Im not very familiar with sp.Popen[bytes] type -> do u mind explaining?

Reply:
[`subprocess.Popen`](https://docs.python.org/3/library/subprocess.html#subprocess.Popen), or as aliased here `sp.Popen`, is the main class that is used open subprocesses in python. It basically works just like [`popen`](https://man7.org/linux/man-pages/man3/popen.3.html) from C.

But `sp.Popen` is actually a generic class in its implementation. Depending on how you initialize the class, you end up with an `sp.Popen[str]` or an `sp.Popen[bytes]`. For all intents and purposes, this basically just says that when you want to send data from the main process to the subprocess or vice-versa (think `stdin`, `stdout`, `stderr`), should you try to communicate using `bytes`s or `str`s.

The "default" way of opening a subprocess is with file objects open in binary mode, meaning that you end up with a `sp.Popen[bytes]`, but you change this in a couple of ways. For e.g. if you open a subprocess with an `text=True` parameter, you'll get a `sp.Popen[str]` (decoding to utf-8 by default iirc)
```py
>>> from subprocess import Popen, PIPE
>>> p = Popen(["echo", "spam", "eggs"], stdout=PIPE)
>>> b ,= p.stdout.readlines()
>>> type(b)
<class 'bytes'>
>>> b
b'spam eggs\n'
>>> p2 = Popen(["echo", "spam", "eggs"], stdout=PIPE, text=True)
>>> s ,= p2.stdout.readlines()
>>> type(s)
<class 'str'>
>>> s
'spam eggs\n'
```
This type hint is just say that I want a dictionary of unique mapping to processes, and if I try to communicate with those processes, I should encode all of my text to `bytes`. I kinda selected this arbitrarily, and am more than willing to be argued into `sp.Popen[str]` or even `sp.Popen[Any]`!!